source/modulo_new_core/include/modulo_new_core/communication/MessagePair.hpp

```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include "modulo_new_core/communication/MessagePairInterface.hpp"
+
+#include <std_msgs/msg/bool.hpp>
+#include <std_msgs/msg/float64.hpp>
+#include <std_msgs/msg/float64_multi_array.hpp>
+#include <std_msgs/msg/int64.hpp>
+#include <std_msgs/msg/string.hpp>
+
+namespace modulo_new_core::communication {
+
+template<typename MsgT, typename DataT>
+class MessagePair : public MessagePairInterface {
+private:
+  std::shared_ptr<DataT> data_;
+
+public:
+  explicit MessagePair(std::shared_ptr<DataT> data);
+
+  [[nodiscard]] MsgT get_message() const;
+
+  [[nodiscard]] std::shared_ptr<DataT> get_data() const;
+};
+
+template<typename MsgT, typename DataT>
+MsgT MessagePair<MsgT, DataT>::get_message() const {
```

Comment:
Ultimately, the `translators` namespace should have corresponding `write_msg(MsgT&, const DataT&)` templates or overloads that you simply call here as `write_msg(msg, *this->data_)`. This avoids any translation logic from being implemented at this layer. Maybe you can mark that with a TODO comment so it's clear.